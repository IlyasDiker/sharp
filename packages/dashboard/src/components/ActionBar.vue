<template>
    <div class="my-4">
        <div class="row gx-3">
            <div class="col">
                <template v-if="filters.length > 0">
                    <div class="row gx-2">
                        <template v-for="filter in filters" :key="filter.id">
                            <div class="col-auto">
                                <SharpFilter
                                    :filter="filter"
                                    :value="filterValue(filter.key)"
                                    @input="$emit('filter-change', filter, $event)"
                                />
                            </div>
                        </template>
                        <template v-if="showReset">
                            <div class="col-auto d-flex">
                                <button class="btn btn-sm btn-link d-inline-flex align-items-center fs-8" @click="$emit('filters-reset', filters)">
                                    {{ l('filters.reset_all') }}
                                </button>
                            </div>
                        </template>
                    </div>
                </template>
            </div>
            <template v-if="commands.length">
                <div class="col-auto">
                    <CommandsDropdown :commands="commands" @select="$emit('command', $event)">
                        <template v-slot:text>
                            {{ l('dashboard.commands.dashboard.label') }}
                        </template>
                    </CommandsDropdown>
                </div>
            </template>
        </div>
    </div>
</template>

<script>
    import { mapGetters } from 'vuex';
    import { SharpFilter } from '@sharp/filters';
    import { CommandsDropdown } from '@sharp/commands';
    import { Localization } from "sharp/mixins";

    export default {
        name: 'SharpActionBarDashboard',
        mixins: [Localization],
        components: {
            SharpFilter,
            CommandsDropdown,
        },
        props: {
            commands: Array,
            filters: Array,
            showReset: Boolean,
        },
        computed: {
            ...mapGetters('dashboard', {
                filterValue: 'filters/value',
            })
        },
    }
</script>
