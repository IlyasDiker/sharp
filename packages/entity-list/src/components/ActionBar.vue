<script setup lang="ts">
    import { __ } from "@/utils/i18n";
    import { config } from "@/utils/config";
</script>

<template>
    <div class="action-bar mt-4 mb-3 position-sticky ShowEntityListField__action-bar"
        v-sticky
        @stuck-change="stuck = $event.detail"
    >
        <div class="row align-items-center">
            <div class="col position-relative">
                <template v-if="config('sharp.display_breadcrumb')">
                    <EntityListTitle :count="count">
                        <Breadcrumb :items="breadcrumb" />
                    </EntityListTitle>
                </template>
            </div>
            <div class="col-auto position-relative">
                <div class="row justify-content-end flex-nowrap gx-3">
                    <template v-if="canReorder && !selecting">
                        <template v-if="reordering">
                            <div class="col-auto">
                                <Button outline @click="handleReorderButtonClicked">
                                    {{ __('sharp::action_bar.list.reorder_button.cancel') }}
                                </Button>
                            </div>
                            <div class="col-auto">
                                <Button @click="handleReorderSubmitButtonClicked">
                                    {{ __('sharp::action_bar.list.reorder_button.finish') }}
                                </Button>
                            </div>
                        </template>
                        <template v-else>
                            <div class="col-auto">
                                <Button outline @click="handleReorderButtonClicked">
                                    {{ __('sharp::action_bar.list.reorder_button') }}
                                </Button>
                            </div>
                        </template>
                    </template>

                    <template v-if="canSelect && !reordering">
                        <template v-if="selecting">
                            <div class="col-auto">
                                <Button key="cancel" outline @click="handleSelectCancelled">
                                    {{ __('sharp::action_bar.list.reorder_button.cancel') }}
                                </Button>
                            </div>
                        </template>
                        <template v-else>
                            <div class="col-auto">
                                <Button key="select" outline @click="handleSelectButtonClicked">
                                    {{ __('sharp::action_bar.list.select_button') }}
                                </Button>
                            </div>
                        </template>
                    </template>

                    <template v-if="hasDropdownCommands && !reordering">
                        <div class="col-auto">
                            <CommandsDropdown
                                class="bg-white"
                                :small="false"
                                :outline="!selecting"
                                :commands="commands"
                                :disabled="reordering"
                                :selecting="selecting"
                                @select="handleCommandSelected"
                            >
                                <template v-slot:text>
                                    {{ __('sharp::entity_list.commands.entity.label') }}
                                    <template v-if="selecting">
                                        ({{ selectedCount }} selected)
                                    </template>
                                </template>
                            </CommandsDropdown>
                        </div>
                    </template>

                    <template v-if="primaryCommand && !reordering && !selecting">
                        <div class="col-auto">
                            <Button @click="handlePrimaryCommandClicked">
                                {{ primaryCommand.label }}
                            </Button>
                        </div>
                    </template>

                    <template v-if="canCreate && !reordering && !selecting">
                        <div class="col-auto">
                            <template v-if="hasForms">
                                <MultiformDropdown
                                    :forms="forms"
                                    right
                                    @select="handleCreateFormSelected"
                                />
                            </template>
                            <template v-else>
                                <Button :disabled="reordering || selecting" @click="handleCreateButtonClicked">
                                    {{ __('sharp::action_bar.list.create_button') }}
                                </Button>
                            </template>
                        </div>
                    </template>
                </div>
            </div>
        </div>
    </div>
</template>

<script lang="ts">
    import { Dropdown,  DropdownItem, Search, Button, Breadcrumb } from '@sharp/ui';
    import { SharpFilter } from '@sharp/filters';

    import MultiformDropdown from "./MultiformDropdown.vue";
    import EntityListTitle from "./EntityListTitle.vue";
    import { CommandsDropdown } from "@sharp/commands";
    import { sticky } from "sharp/directives";

    export default {
        name: 'SharpActionBarList',

        components : {
            CommandsDropdown,
            EntityListTitle,
            MultiformDropdown,
            Dropdown,
            DropdownItem,
            SharpFilter,
            Search,
            Button,
            Breadcrumb,
        },

        props: {
            count: Number,
            search: String,
            filters: Array,
            filtersValues: Object,
            forms: Array,
            commands: Array,

            canCreate: Boolean,
            canReorder: Boolean,
            canSearch: Boolean,
            canSelect: Boolean,

            reordering: Boolean,
            selecting: Boolean,
            selectedCount: Number,

            breadcrumb: Array,
        },
        data() {
            return {
                stuck: false,
            }
        },
        computed: {
            hasDropdownCommands() {
                return this.commands?.flat().filter(command => !command.primary).length > 0;
            },
            primaryCommand() {
                return this.commands?.flat().find(command => command.primary);
            },
            hasForms() {
                return this.forms && this.forms.length > 0;
            },
        },
        methods: {
            handleSearchSubmitted(search) {
                this.$emit('search-submit', search);
            },
            handleFilterChanged(filter, value) {
                this.$emit('filter-change', filter, value);
            },
            handlePrimaryCommandClicked() {
                this.$emit('command', this.primaryCommand);
            },
            handleReorderButtonClicked() {
                this.$emit('reorder-click');
                document.activeElement.blur();
            },
            handleReorderSubmitButtonClicked() {
                this.$emit('reorder-submit');
            },
            handleCreateButtonClicked() {
                this.$emit('create');
            },
            handleCreateFormSelected(form) {
                this.$emit('create', form);
            },
            handleCommandSelected(command) {
                this.$emit('command', command);
            },
            handleSelectButtonClicked() {
                this.$emit('select-click');
            },
            handleSelectCancelled() {
                this.$emit('select-cancel');
            },
        },
        directives: {
            sticky
        },
    }
</script>
