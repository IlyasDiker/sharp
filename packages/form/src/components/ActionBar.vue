<script setup lang="ts">
    import { Breadcrumb } from '@sharp/ui';
    import LocaleSelect from "./ui/LocaleSelect.vue";
    import { BreadcrumbData } from "@/types";
    import { config } from "@/utils/config";

    defineProps<{
        breadcrumb: BreadcrumbData,
        locales?: Array<string>,
    }>();
</script>

<template>
    <div class="action-bar mt-4 mb-3">
        <div class="row align-items-center gx-3">
            <div class="col">
                <template v-if="config('sharp.display_breadcrumb')">
                    <Breadcrumb :items="breadcrumb.items" />
                </template>
            </div>
            <template v-if="locales?.length">
                <div class="col-auto">
                    <LocaleSelect
                        outline
                        right
                        :locale="currentLocale"
                        :locales="locales"
                        @change="$emit('locale-change', $event)"
                    />
                </div>
            </template>
        </div>
    </div>
</template>
