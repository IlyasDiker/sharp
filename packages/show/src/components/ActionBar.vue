<template>
    <ActionBar>
        <template v-slot:left>
            <div class="row align-items-center gx-4">
                <template v-if="showBackButton">
                    <div class="col-auto">
                        <Button :href="backUrl" outline variant="light" large>
                            {{ l('action_bar.show.back_button') }}
                        </Button>
                    </div>
                </template>
                <template v-if="title">
                    <div class="col">
                        <div class="h5 mb-0" :class="{ 'opacity-0': !showTitle }" style="transition: opacity .2s ease-in-out">
                            {{ title }}
                        </div>
                    </div>
                </template>
            </div>

        </template>
        <template v-slot:right>
            <template v-if="canEdit">
                <Button :href="formUrl" :disabled="editDisabled" variant="light" large>
                    {{ l('action_bar.show.edit_button') }}
                </Button>
            </template>
        </template>
        <template v-slot:extras>
            <template v-if="showBreadcrumb">
                <Breadcrumb :items="breadcrumb" />
            </template>
        </template>
        <template v-slot:extras-right>
            <div class="row gx-3">
                <template v-if="hasState">
                    <div class="col-auto">
                        <ModalSelect
                            :title="l('modals.entity_state.edit.title')"
                            :ok-title="l('modals.entity_state.edit.ok_button')"
                            :value="state.value"
                            :options="stateValues"
                            size="sm"
                            @change="handleStateChanged"
                        >
                            <template v-slot="{ on }">
                                <Button
                                    class="btn--opacity-1"
                                    :class="{ 'dropdown-toggle':canChangeState }"
                                    :disabled="!canChangeState"
                                    text
                                    small
                                    v-on="on"
                                >
                                    <StateIcon class="me-1" :color="state.color" style="vertical-align: -.125em" />
                                    <span class="text-truncate">{{ state.label }}</span>
                                </Button>
                            </template>

                            <template v-slot:item-prepend="{ option }">
                                <StateIcon :color="option.color" />
                            </template>
                        </ModalSelect>
                    </div>
                </template>
                <template v-if="hasCommands">
                    <div class="col-auto">
                        <CommandsDropdown :commands="commands" @select="handleCommandSelected">
                            <template v-slot:text>
                                {{ l('entity_list.commands.instance.label') }}
                            </template>
                        </CommandsDropdown>
                    </div>
                </template>
            </div>
        </template>
    </ActionBar>
</template>

<script>
    import {
        ActionBar,
        Dropdown,
        DropdownItem,
        StateIcon,
        Breadcrumb,
        Button,
        ModalSelect,
    } from 'sharp-ui';

    import {
        CommandsDropdown
    } from 'sharp-commands';

    import { Localization } from "sharp/mixins";

    export default {
        mixins: [Localization],
        components: {
            ActionBar,
            Breadcrumb,
            CommandsDropdown,
            Dropdown,
            DropdownItem,
            StateIcon,
            Button,
            ModalSelect,
        },
        props: {
            commands: Array,
            formUrl: String,
            backUrl: String,
            canEdit: Boolean,
            editDisabled: Boolean,
            canChangeState: Boolean,
            showBackButton: Boolean,
            state: Object,
            stateValues: Array,
            breadcrumb: Array,
            showBreadcrumb: Boolean,
        },
        data() {
            return {
                showTitle: false,
            }
        },
        computed: {
            hasCommands() {
                return this.commands?.flat().length > 0;
            },
            hasState() {
                return !!this.state;
            },
            title() {
                return this.breadcrumb && this.showBreadcrumb
                    ? this.breadcrumb[this.breadcrumb.length - 1]?.name
                    : null;
            },
        },
        methods: {
            handleEditButtonClicked() {
                this.$emit('edit');
            },
            handleCommandSelected(command) {
                this.$emit('command', command);
            },
            handleStateChanged(state) {
                this.$emit('state-change', state);
            },
            handleScroll() {
                this.showTitle = document.querySelector('.ShowPage__content').getBoundingClientRect().top < 0;
            },
        },
        mounted() {
            window.addEventListener('scroll', this.handleScroll);
        },
    }
</script>
