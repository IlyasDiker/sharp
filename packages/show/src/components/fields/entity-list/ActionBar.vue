<script setup lang="ts">
    import { __ } from "@/utils/i18n";
</script>

<template>
    <div class="action-bar"
        :class="{ 'position-sticky': sticky }"
        v-sticky="sticky"
        @stuck-change="stuck = $event.detail"
    >
        <div class="position-relative">
            <template v-if="hasOuterTitle">
                <div class="mb-2">
                    <slot />
                </div>
            </template>
            <template v-if="ready && barVisible">
                <div class="row align-items-end align-content-end" data-sticky-anchor>
                    <div class="col-sm align-self-end">
                        <slot />
                    </div>

                    <template v-if="hasRightControls && !collapsed">
                        <div class="col-sm-auto mb-2">
                            <div class="row flex-nowrap justify-content-end g-2 gx-md-3">
                                <template v-if="canReorder">
                                    <div class="col-auto">
                                        <template v-if="reordering">
                                            <div class="row gx-3">
                                                <div class="col-auto">
                                                    <Button text small @click="handleReorderButtonClicked">
                                                        {{ __('sharp::action_bar.list.reorder_button.cancel') }}
                                                    </Button>
                                                </div>
                                                <div class="col-auto">
                                                    <Button variant="primary" small @click="handleReorderSubmitButtonClicked">
                                                        {{ __('sharp::action_bar.list.reorder_button.finish') }}
                                                    </Button>
                                                </div>
                                            </div>
                                        </template>
                                        <template v-else>
                                            <Button text small @click="handleReorderButtonClicked">
                                                {{ __('sharp::action_bar.list.reorder_button') }}
                                            </Button>
                                        </template>
                                    </div>
                                </template>

                                <template v-if="primaryCommand && !reordering">
                                    <div class="col-auto">
                                        <Button variant="primary" small @click="handlePrimaryCommandClicked">
                                            {{ primaryCommand.label }}
                                        </Button>
                                    </div>
                                </template>

                                <template v-if="canCreate && !reordering">
                                    <div class="col-auto">
                                        <div class="action-bar__element">
                                            <template v-if="hasForms">
                                                <MultiformDropdown
                                                    :forms="forms"
                                                    variant="primary"
                                                    right
                                                    small
                                                    @select="handleCreateFormSelected"
                                                />
                                            </template>
                                            <template v-else>
                                                <Button variant="primary" small @click="handleCreateButtonClicked">
                                                    {{ __('sharp::action_bar.list.create_button') }}
                                                </Button>
                                            </template>
                                        </div>
                                    </div>
                                </template>
                            </div>
                        </div>
                    </template>
                </div>
            </template>
        </div>
    </div>
</template>

<script lang="ts">
    import { Button } from '@sharp/ui';
    import { MultiformDropdown } from "@sharp/entity-list";
    import { sticky } from "sharp/directives";

    export default {
        components: {
            MultiformDropdown,
            Button,
        },
        props: {
            ready: Boolean,
            count: Number,
            primaryCommand: Object,

            forms: Array,

            canCreate: Boolean,
            canReorder: Boolean,

            reordering: Boolean,
            selecting: Boolean,

            // show field props
            collapsed: Boolean,
            hasActiveQuery: Boolean,
            sticky: Boolean,
        },
        data() {
            return {
                stuck: false,
            }
        },
        computed: {
            hasForms() {
                return this.forms && this.forms.length > 0;
            },
            hasLeftControls() {
                return false;
                // return this.hasActiveQuery || this.count > 0 && (this.filters?.length > 0 || this.canSearch);
            },
            hasRightControls() {
                return this.canReorder || this.canCreate || !!this.primaryCommand;
            },
            hasOuterTitle() {
                return this.$slots.default && (!this.ready || this.hasLeftControls);
            },
            barVisible() {
                if(this.collapsed) {
                    return !this.hasLeftControls;
                }
                return true;
            },
        },
        methods: {
            handleFilterChanged(filter, value) {
                this.$emit('filter-change', filter, value);
            },
            handleReorderButtonClicked() {
                this.$emit('reorder-click');
                document.activeElement.blur();
            },
            handleReorderSubmitButtonClicked() {
                this.$emit('reorder-submit');
            },
            handlePrimaryCommandClicked() {
                this.$emit('command', this.primaryCommand);
            },
            handleCreateButtonClicked() {
                this.$emit('create');
            },
            handleCreateFormSelected(form) {
                this.$emit('create', form);
            },
        },
        directives: {
            sticky,
        },
    }
</script>
