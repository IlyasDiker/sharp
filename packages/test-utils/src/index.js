export { default as MockInjections } from './MockInjections.vue';
export { default as MockI18n } from './MockI18n.vue';
export { default as QueryComponent } from './QueryComponent.vue';
export { default as MockTransitions } from './MockTransitions.vue';
export { mockSFC, unmockSFC } from './mockSFC.vue';
export { mockProperty, unmockProperty, setter } from './mock-utils';
export { nextRequestFulfilled } from './moxios-utils';
export { createStub } from './stubs';

export function wait(delay) {
    return new Promise(resolve => {
        setTimeout(resolve, delay);
    });
}
