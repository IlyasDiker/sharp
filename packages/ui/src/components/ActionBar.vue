<template>
    <div class="SharpActionBar">
        <div class="SharpActionBar__bar sticky-top" ref="bar">
            <div class="container">
                <div class="row align-items-center g-2 gx-md-4">
                    <div class="col SharpActionBar__col--left" style="min-width: 0">
                        <slot name="left"></slot>
                    </div>
                    <div class="col-auto SharpActionBar__col--right">
                        <slot name="right"></slot>
                    </div>
                </div>
            </div>
        </div>
        <template v-if="hasExtras">
            <div :class="{ 'container':container }">
                <div class="row">
                    <div class="col-sm">
                        <div class="SharpActionBar__extras">
                            <slot name="extras" />
                        </div>
                    </div>
                    <template v-if="$slots['extras-right']">
                        <div class="col-sm-auto">
                            <div class="SharpActionBar__extras">
                                <slot name="extras-right" />
                            </div>
                        </div>
                    </template>
                </div>
            </div>
        </template>
    </div>
</template>

<script>
    import { getNavbarHeight } from "../util";

    export default {
        name: 'SharpActionBar',
        props: {
            container: Boolean,
        },
        computed: {
            hasExtras() {
                return this.$slots.extras || this.$slots['extras-right'];
            },
        },
        methods: {
            layout(rect) {
                /** @see getNavbarHeight **/
                document.documentElement.style.setProperty('--navbar-height', `${rect.height}px`);
            },
        },
        mounted() {
            this.layout(this.$refs.bar.getBoundingClientRect());

            if(window.ResizeObserver) {
                (new ResizeObserver(entries => {
                    this.layout(entries[0].target.getBoundingClientRect());
                }))
                .observe(this.$refs.bar);
            }
        }
    }
</script>
