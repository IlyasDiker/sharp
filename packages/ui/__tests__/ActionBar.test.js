import { mount } from '@vue/test-utils';
import ActionBar from '../src/components/TopBar.vue';

describe('action-bar', ()=>{
    test('can mount ActionBar', ()=>{
        expect(mount(ActionBar).html()).toMatchSnapshot();
    });
    test('can mount with slots ActionBar', ()=>{
        expect(mount(ActionBar, {
            slots: {
                left: '<div>LEFT CONTENT</div>',
                right: '<div>RIGHT CONTENT</div>',
                extras: '<div>EXTRAS CONTENT</div>'
            }
        }).html()).toMatchSnapshot();
    });
});
