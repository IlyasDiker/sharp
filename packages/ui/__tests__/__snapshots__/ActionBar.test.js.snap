// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`action-bar can mount ActionBar 1`] = `
<div class="SharpActionBar">
	<div class="SharpActionBar__bar sticky-top">
		<div class="container">
			<div class="row align-items-center g-2 gx-md-4">
				<div class="col SharpActionBar__col--left" style="min-width: 0;"></div>
				<div class="col-auto SharpActionBar__col--right"></div>
			</div>
		</div>
	</div>
	<!---->
</div>
`;

exports[`action-bar can mount with slots ActionBar 1`] = `
<div class="SharpActionBar">
	<div class="SharpActionBar__bar sticky-top">
		<div class="container">
			<div class="row align-items-center g-2 gx-md-4">
				<div class="col SharpActionBar__col--left" style="min-width: 0;">
					<div>LEFT CONTENT</div>
				</div>
				<div class="col-auto SharpActionBar__col--right">
					<div>RIGHT CONTENT</div>
				</div>
			</div>
		</div>
	</div>
	<div class="">
		<div class="row">
			<div class="col-sm">
				<div class="SharpActionBar__extras">
					<div>EXTRAS CONTENT</div>
				</div>
			</div>
			<!---->
		</div>
	</div>
</div>
`;
